// @deck.gl/core/dist/es5/effects/mask/mask-effect.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _core = require("@luma.gl/core");

var _core2 = require("@math.gl/core");

var _maskPass = _interopRequireDefault(require("../../passes/mask-pass"));

var _effect = _interopRequireDefault(require("../../lib/effect"));

var _constants = require("../../lib/constants");

var _utils = require("./utils");

var _log = _interopRequireDefault(require("../../utils/log"));

function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it.return != null) it.return(); } finally { if (didErr) throw err; } } }; }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var MaskEffect = function (_Effect) {
  (0, _inherits2.default)(MaskEffect, _Effect);

  var _super = _createSuper(MaskEffect);

  function MaskEffect(props) {
    var _this;

    (0, _classCallCheck2.default)(this, MaskEffect);
    _this = _super.call(this, props);
    _this.dummyMaskMap = null;
    _this.useInPicking = true;
    _this.channels = [];
    _this.masks = null;
    return _this;
  }

  (0, _createClass2.default)(MaskEffect, [{
    key: "preRender",
    value: function preRender(gl, _ref) {
      var layers = _ref.layers,
          layerFilter = _ref.layerFilter,
          viewports = _ref.viewports,
          onViewportActive = _ref.onViewportActive,
          views = _ref.views;

      if (!this.dummyMaskMap) {
        this.dummyMaskMap = new _core.Texture2D(gl, {
          width: 1,
          height: 1
        });
      }

      var maskLayers = layers.filter(function (l) {
        return l.props.operation === _constants.OPERATION.MASK && l.props.visible;
      });

      if (maskLayers.length === 0) {
        this.masks = null;
        this.channels.length = 0;
        return;
      }

      this.masks = {};

      if (!this.maskPass) {
        this.maskPass = new _maskPass.default(gl, {
          id: 'default-mask'
        });
        this.maskMap = this.maskPass.maskMap;
      }

      var channelMap = this._sortMaskChannels(maskLayers);

      var viewport = viewports[0];
      var viewportChanged = !this.lastViewport || !this.lastViewport.equals(viewport);

      for (var maskId in channelMap) {
        this._renderChannel(channelMap[maskId], {
          layerFilter: layerFilter,
          onViewportActive: onViewportActive,
          views: views,
          viewport: viewport,
          viewportChanged: viewportChanged
        });
      }
    }
  }, {
    key: "_renderChannel",
    value: function _renderChannel(channelInfo, _ref2) {
      var layerFilter = _ref2.layerFilter,
          onViewportActive = _ref2.onViewportActive,
          views = _ref2.views,
          viewport = _ref2.viewport,
          viewportChanged = _ref2.viewportChanged;
      var oldChannelInfo = this.channels[channelInfo.index];
      var maskChanged = channelInfo === oldChannelInfo || oldChannelInfo.layers.length !== channelInfo.layers.length || channelInfo.layerBounds.some(function (b, i) {
        return b !== oldChannelInfo.layerBounds[i];
      });
      channelInfo.bounds = oldChannelInfo.bounds;
      channelInfo.maskBounds = oldChannelInfo.maskBounds;
      this.channels[channelInfo.index] = channelInfo;

      if (maskChanged || viewportChanged) {
        this.lastViewport = viewport;
        channelInfo.bounds = (0, _utils.getMaskBounds)({
          layers: channelInfo.layers,
          viewport: viewport
        });

        if (maskChanged || !(0, _core2.equals)(channelInfo.bounds, oldChannelInfo.bounds)) {
          var maskPass = this.maskPass,
              maskMap = this.maskMap;
          var maskViewport = (0, _utils.getMaskViewport)({
            bounds: channelInfo.bounds,
            viewport: viewport,
            width: maskMap.width,
            height: maskMap.height
          });
          channelInfo.maskBounds = maskViewport ? maskViewport.getBounds() : [0, 0, 1, 1];
          maskPass.render({
            channel: channelInfo.index,
            layers: channelInfo.layers,
            layerFilter: layerFilter,
            viewports: maskViewport ? [maskViewport] : [],
            onViewportActive: onViewportActive,
            views: views,
            moduleParameters: {
              devicePixelRatio: 1
            }
          });
        }
      }

      this.masks[channelInfo.id] = {
        index: channelInfo.index,
        bounds: channelInfo.maskBounds,
        coordinateOrigin: channelInfo.coordinateOrigin,
        coordinateSystem: channelInfo.coordinateSystem
      };
    }
  }, {
    key: "_sortMaskChannels",
    value: function _sortMaskChannels(maskLayers) {
      var _this2 = this;

      var channelMap = {};
      var channelCount = 0;

      var _iterator = _createForOfIteratorHelper(maskLayers),
          _step;

      try {
        var _loop = function _loop() {
          var layer = _step.value;
          var id = layer.root.id;
          var channelInfo = channelMap[id];

          if (!channelInfo) {
            if (++channelCount > 4) {
              _log.default.warn('Too many mask layers. The max supported is 4')();

              return "continue";
            }

            channelInfo = {
              id: id,
              index: _this2.channels.findIndex(function (c) {
                return (c === null || c === void 0 ? void 0 : c.id) === id;
              }),
              layers: [],
              layerBounds: [],
              coordinateOrigin: layer.root.props.coordinateOrigin,
              coordinateSystem: layer.root.props.coordinateSystem
            };
            channelMap[id] = channelInfo;
          }

          channelInfo.layers.push(layer);
          channelInfo.layerBounds.push(layer.getBounds());
        };

        for (_iterator.s(); !(_step = _iterator.n()).done;) {
          var _ret = _loop();

          if (_ret === "continue") continue;
        }
      } catch (err) {
        _iterator.e(err);
      } finally {
        _iterator.f();
      }

      for (var i = 0; i < 4; i++) {
        var channelInfo = this.channels[i];

        if (!channelInfo || !(channelInfo.id in channelMap)) {
          this.channels[i] = null;
        }
      }

      for (var maskId in channelMap) {
        var _channelInfo = channelMap[maskId];

        if (_channelInfo.index < 0) {
          _channelInfo.index = this.channels.findIndex(function (c) {
            return !c;
          });
          this.channels[_channelInfo.index] = _channelInfo;
        }
      }

      return channelMap;
    }
  }, {
    key: "getModuleParameters",
    value: function getModuleParameters() {
      return {
        maskMap: this.masks ? this.maskMap : this.dummyMaskMap,
        maskChannels: this.masks
      };
    }
  }, {
    key: "cleanup",
    value: function cleanup() {
      if (this.dummyMaskMap) {
        this.dummyMaskMap.delete();
        this.dummyMaskMap = null;
      }

      if (this.maskPass) {
        this.maskPass.delete();
        this.maskPass = null;
        this.maskMap = null;
      }

      this.lastViewport = null;
      this.masks = null;
      this.channels.length = 0;
    }
  }]);
  return MaskEffect;
}(_effect.default);

exports.default = MaskEffect;
