// @loaders.gl/textures/dist/es5/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "BasisLoader", {
  enumerable: true,
  get: function get() {
    return _basisLoader.BasisLoader;
  }
});
Object.defineProperty(exports, "BasisWorkerLoader", {
  enumerable: true,
  get: function get() {
    return _basisLoader.BasisWorkerLoader;
  }
});
Object.defineProperty(exports, "CompressedTextureLoader", {
  enumerable: true,
  get: function get() {
    return _compressedTextureLoader.CompressedTextureLoader;
  }
});
Object.defineProperty(exports, "CompressedTextureWorkerLoader", {
  enumerable: true,
  get: function get() {
    return _compressedTextureLoader.CompressedTextureWorkerLoader;
  }
});
Object.defineProperty(exports, "CrunchLoader", {
  enumerable: true,
  get: function get() {
    return _crunchLoader.CrunchLoader;
  }
});
Object.defineProperty(exports, "CrunchWorkerLoader", {
  enumerable: true,
  get: function get() {
    return _crunchLoader.CrunchLoader;
  }
});
Object.defineProperty(exports, "NPYLoader", {
  enumerable: true,
  get: function get() {
    return _npyLoader.NPYLoader;
  }
});
Object.defineProperty(exports, "NPYWorkerLoader", {
  enumerable: true,
  get: function get() {
    return _npyLoader.NPYWorkerLoader;
  }
});
Object.defineProperty(exports, "CompressedTextureWriter", {
  enumerable: true,
  get: function get() {
    return _compressedTextureWriter.CompressedTextureWriter;
  }
});
Object.defineProperty(exports, "KTX2BasisUniversalTextureWriter", {
  enumerable: true,
  get: function get() {
    return _ktx2BasisUniversalTextureWriter.KTX2BasisUniversalTextureWriter;
  }
});
Object.defineProperty(exports, "loadImageTexture", {
  enumerable: true,
  get: function get() {
    return _loadImage.loadImageTexture;
  }
});
Object.defineProperty(exports, "loadImageTextureArray", {
  enumerable: true,
  get: function get() {
    return _loadImageArray.loadImageTextureArray;
  }
});
Object.defineProperty(exports, "loadImageTextureCube", {
  enumerable: true,
  get: function get() {
    return _loadImageCube.loadImageTextureCube;
  }
});
Object.defineProperty(exports, "GL_EXTENSIONS_CONSTANTS", {
  enumerable: true,
  get: function get() {
    return _glExtensions.GL_EXTENSIONS_CONSTANTS;
  }
});
Object.defineProperty(exports, "selectSupportedBasisFormat", {
  enumerable: true,
  get: function get() {
    return _parseBasis.selectSupportedBasisFormat;
  }
});
Object.defineProperty(exports, "getSupportedGPUTextureFormats", {
  enumerable: true,
  get: function get() {
    return _textureFormats.getSupportedGPUTextureFormats;
  }
});

var _basisLoader = require("./basis-loader");

var _compressedTextureLoader = require("./compressed-texture-loader");

var _crunchLoader = require("./crunch-loader");

var _npyLoader = require("./npy-loader");

var _compressedTextureWriter = require("./compressed-texture-writer");

var _ktx2BasisUniversalTextureWriter = require("./ktx2-basis-universal-texture-writer");

var _loadImage = require("./lib/texture-api/load-image");

var _loadImageArray = require("./lib/texture-api/load-image-array");

var _loadImageCube = require("./lib/texture-api/load-image-cube");

var _glExtensions = require("./lib/gl-extensions");

var _parseBasis = require("./lib/parsers/parse-basis");

var _textureFormats = require("./lib/utils/texture-formats");
